apartmenttherapy/oxygen-lib-ui
# add new placeholders pt 2

#### What does this PR do & why?

This PR adjusts the placeholders for in-post images and galleries, from a chunky pixelated preview of the original image to a solid ciolor (gray) image. The motivation is to save precious bytes on initial load of the page.

Doing this required that we support multiple _transforms_ (array of `renderOps`) per image. That was actually pretty easy; I made it so the `renderOps` arg to `createUrl` can accept either an object or array of objects.

This is follow-up to #241 

#### How can this be manually tested?

`yarn dev` and check it out on a post with lots of images and a gallery, like http://localhost:8080/shrimp-boil-recipe-259082

you also might want to check it on AT via `SITE=at yarn dev` and this page: http://localhost:8080/phoebe-robinson-brookyn-nyc-apartment-photos-263152

also, make sure @chenryhen recent work #240 isn't broken.

#### Any questions or concerns about this change?

I think the code is pretty solid. Adding support for a second transform was "interesting".

The only real tricky/annoying part was when we merge `renderOps` and `BASE_RENDER_OPS`: we only want to apply the `BASE_RENDER_OPS` to the _last_ transform. Same goes for `Figure`, where we copy the width/height. If `Figure` used  our `Image` component, life would be somewhat easier since I just refactored things so that `Image` can handle rendering a placeholder image.

also, see https://gist.github.com/blech75/ec3565997bcf832d94d9b2ffc0f92c40 for some timing measurements on first render of image vs. cached for old-style pixelated vs. new-style single-color.


#### Deployment Notes

nope.

#### Link to story card and other documentation

see https://app.asana.com/0/543787180217047/930017999839002/f

#### GIF

![https://media.giphy.com/media/HOGGwOO2kduXC/giphy.gif](https://media.giphy.com/media/HOGGwOO2kduXC/giphy.gif)

#### Have you remembered to...?
